---
title: "Differential Expression: RNA-seq example"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
params:
  baseDir: ""
  annotation_file: ""
  design: !r formula("~ Timepoint")
  contrasts: !r list(c(factor = "Timepoint",
        numerator = "90",
        denominator = "0"))
  alpha: 0.1
  lfc_threshold: 0
  data_dir: !r file.path("../data", Sys.Date())
  results_dir: !r file.path("../results", Sys.Date(), "differential_expression")
editor_options:
  chunk_output_type: console
output:
  html_document:
    code_folding: "hide"
---

```{r setup, message=FALSE, warning=FALSE}
args = commandArgs(trailingOnly=TRUE)

options(stringsAsFactors = FALSE)
bcbioRNASeq::prepareRNASeqTemplate()
source("_setup.R"); source("_utils.R"); source("_plotting.R")

library(data.table); library(pca3d); library(rcartocolor); library(genefilter); library(tximport)

# Directory paths ==============================================================
invisible(mapply(
    FUN = dir.create,
    path = c(params$data_dir, params$results_dir),
    MoreArgs = list(showWarnings = FALSE, recursive = TRUE)
))

```



```{r import_counts, message=FALSE, warning=FALSE, results='hide'}

# Read in gencode annotations
gtf <- rtracklayer::import("/home/msmallegan/rmghc-workshop-19/results/genome/gencode.vM21.annotation.gtf")
#gtf <- rtracklayer::import(params$annotation_file)



#TODO: put this parsing in a setup file to clean this up.
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_id")] <-
  "geneID"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_name")] <-
  "geneName"
names(gtf@elementMetadata@listData)[which(names(gtf@elementMetadata@listData) == "gene_type")] <-
  "geneBiotype"
gtf@elementMetadata$id <- gtf@elementMetadata$geneID


gtfdf <- gtf %>% as.data.frame()
g2s <- as.data.frame(gtf@elementMetadata@listData) %>% select(geneID, geneName) %>% distinct()

tx2gene <- gtfdf %>%
  mutate(TXNAME = transcript_id, GENEID = geneID) %>%
  select(TXNAME, GENEID)



# Read in sample annotations
samples <- read.csv("sample_sheet.csv") %>%
  mutate(Timepoint = factor(Timepoint,
         levels = c("0", "90")))



# Read in the salmon counts data.
files <- file.path(samples$SampleID, "quant.sf")
names(files) <- samples$SampleID
txi.salmon <- tximport(files, type = "salmon", tx2gene = tx2gene)


# Ensure that the ordering of the columns in the counts matrix
# is the same as in the sample sheet.
rownames(samples) <- samples$SampleID
samples <- samples[colnames(txi.salmon$counts),]
stopifnot(all(rownames(samples) == colnames(txi.salmon$counts)))

# Ensure that the ordering of the rows (gene info)
# is the same between the counts and the gencode
names(gtf) <- gtf$geneID
gtf <- gtf[rownames(txi.salmon$counts)]
stopifnot(all(names(gtf) == rownames(txi.salmon$counts)))

# Create DESeq2 experiment objects.
dds <- DESeq2::DESeqDataSetFromTximport(txi.salmon,
                                        colData   = samples,
                                        design    = params$design,
                                        rowRanges = gtf)

# Note that this criteria can be made more stringent
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep, ]

dds <- DESeq(dds)
saveData(dds, dir = params$data_dir)

# Alternatively, can use `rlog()` here, but it is slower
rld <- varianceStabilizingTransformation(dds)
interestingGroups(rld) <- c("Timepoint")
saveData(rld, dir = params$data_dir)

# factor; numerator; denominator
# levels(dds$genotype)
# help("results", "DESeq2")
stopifnot(is.list(params$contrasts))
contrasts <- params$contrasts
names <- vapply(
    X = contrasts,
    FUN = function(x) {
        paste(x[[1]], x[[2]], "vs", x[[3]], sep = "_")

    },
    FUN.VALUE = "character"
)
names(contrasts) <- names

res_list_unshrunken <- mapply(
    FUN = results,
    contrast = contrasts,
    MoreArgs = list(
        object = dds,
        alpha = params$alpha
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = FALSE
)
names(res_list_unshrunken) <- names
saveData(res_list_unshrunken, dir = params$data_dir)

# For `type` arguments other than "normal", `coef` argument is required.
# Otherwise can use `contrast`, which is more intuitive and readable.
# If using `coef` number, must match order in `resultsNames(dds)`.
# The new apeglm method is great but currently only works with coef.
# help("lfcShrink", "DESeq2")
# help("coef", "DESeq2")
# help("resultsNames", "DESeq2")
# This step can be a little slow and sped up with the `parallel` argument.
res_list_shrunken <- mapply(
    FUN = lfcShrink,
    res = res_list_unshrunken,
    contrast = contrasts,
    MoreArgs = list(
        dds = dds,
        type = "normal"
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)
saveData(res_list_shrunken, dir = params$data_dir)
```

# Plots

## PCA

Principal Component Analysis (PCA) is a statistical technique used to identify global patterns in high-dimensional datasets. It is commonly used to explore the similarity of biological samples in RNA-seq datasets. To achieve this, gene expression values are transformed into Principal Components (PCs), a set of linearly uncorrelated features which represent the most relevant sources of variance in the data, and subsequently visualized using a scatter plot.

Each point represents an RNA-seq sample. Samples with similar gene expression profiles are closer in the three-dimensional space.

```{r plot_deg_pca, message=FALSE, fig.width=10L, fig.height=6L}
plotDEGPCA(res_list_shrunken[[1]], counts = rld)
```


## Heatmap

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row and used the `ward.D2` method for clustering [@Ward1963-xf].

```{r plot_deg_heatmap}
plotDEGHeatmap(res_list_shrunken[[1]], counts = rld, clusteringMethod = "ward.D2", scale = "row")
```

## Gene scatter plots

### Mean average (MA)
An MA plot compares transformed counts on `M` (log ratio) and `A` (mean average) scales [@Yang2002-sx].

```{r plot_ma}
plotMA(res_list_shrunken[[1]], lfcThreshold = params$lfc_threshold, ntop = 15, gene2symbol = g2s)
```

### Volcano

A volcano plot compares significance (BH-adjusted *P* value) against fold change (log2) [@Cui2003-rn; @Li2014-ll].

```{r plot_volcano}
plotVolcano(res_list_shrunken[[1]], lfcThreshold = params$lfc_threshold, ntop = 15, gene2symbol = g2s)
```
